ExecProbeTimeout feature gate introduction (#24692)
kubernetes/website: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md

* `periodSeconds`: How often (in seconds) to perform the probe. Default to 10
seconds. Minimum value is 1.
* `timeoutSeconds`: Number of seconds after which the probe times out. Defaults
to 1 second. Minimum value is 1. For exec probes, the process inside the container

> may keep running even after probe returned failure because of the timeout. Incorrect

Reviewer: Since this only applies for dockershim is it worthwhile to mention that?

Author: as much as I want to migrate everybody out of DockerShim - there are still a lot of customers on DockerShim. Perhaps I can clarify that this only applies to DockerShim. I think this issue is worth mentioning so there would be no surprises.

implementation of readiness probe may result in ever growing number of processes
in the container and resource starvation.
* `successThreshold`: Minimum consecutive successes for the probe to be
considered successful after having failed. Defaults to 1. Must be 1 for
liveness. Minimum value is 1.
* `failureThreshold`: When a probe fails, Kubernetes will
try `failureThreshold` times before giving up. Giving up in case of liveness probe means restarting the container. In case of readiness probe the Pod will be marked Unready.
Defaults to 3. Minimum value is 1.

{{< note >}}

> Before the version `1.20` the field `timeoutSeconds` was not respected for exec probes and probe never returned error when deadline was exceeded. So these probes may run indefinitely.

Reviewer: The probe timeout is now respected, but in dockershim we don't stop the probe from running indefinitely, not sure we need to clarify that.

Author: added clarificaton and shortened the sentence

> Before the version `1.20` the field `timeoutSeconds` was not respected for exec probes and probe never returned error when deadline was exceeded. So these probes may run indefinitely.
> Starting `1.20` this behavior was fixed. However, this change may not be expected. Especially when the default `timeoutSeconds` value of 1 seconds is assumed.
> Cluster administrator may set the feature gate `ExecProbeTimeout` to `false` to restore the previous behavior while the `timeoutSeconds` is not corrected.

Reviewer 2: ```suggestion
Before Kubernetes 1.20, the field `timeoutSeconds` was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.

This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior, even without realizing it, as the default timeout is 1 second.  
As a cluster administrator, you can disable the feature gate `ExecProbeTimeout` (set it to `false`) to restore the  behavior from older versions, then remove that override once all the exec probes in the cluster have a `timeoutSeconds` value set.
```

try `failureThreshold` times before giving up. Giving up in case of liveness probe means restarting the container. In case of readiness probe the Pod will be marked Unready.
Defaults to 3. Minimum value is 1.

{{< note >}}
Before Kubernetes 1.20, the field `timeoutSeconds` was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.

This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior,
even without realizing it, as the default timeout is 1 second.
As a cluster administrator, you can disable the feature gate `ExecProbeTimeout` (set it to `false`)
to restore the  behavior from older versions, then remove that override once all the exec probes
in the cluster have a `timeoutSeconds` value set.

For exec probes, the process inside the container may keep running even after probe returned failure because of the timeout.

> This is a current behavior when DockerShim is used as a runtime on Kubernetes v1.20.

Reviewer 2: > This is a current behavior when DockerShim is used as a runtime on Kubernetes v1.20.

I'm not sure I understand this. Is this also true:
```suggestion
For example, on Kubernetes 1.20 with the `dockershim` container runtime, the probe process
continues running inside the container even after it has timed out.
```

> Incorrect implementation of readiness probe may result in ever growing number of processes
> in the container and resource starvation.
> {{< /note >}}

Reviewer 2: ```suggestion
{{< /note >}}
{{< caution >}}
Incorrect implementation of readiness probes may result in an ever growing number
of processes in the container, and resource starvation if this is left unchecked.
{{< /caution >}}
```
?

try `failureThreshold` times before giving up. Giving up in case of liveness probe means restarting the container. In case of readiness probe the Pod will be marked Unready.
Defaults to 3. Minimum value is 1.

{{< note >}}

> Before Kubernetes 1.20, the field `timeoutSeconds` was not respected for exec probes: probes continued running indefinitely, even past their configured deadline, until a result was returned.

Reviewer: > running indefinitely

This is only true for dockershim, containerd did stop probes within the context timeout we gave it.

Reviewer: Oh I see the section below now, should we remove this bit here then?

This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior,
even without realizing it, as the default timeout is 1 second.

> As a cluster administrator, you can disable the feature gate `ExecProbeTimeout` (set it to `false`)

Reviewer: Probably worth calling out that this is only required in the kubelet

Author: added

to restore the  behavior from older versions, then remove that override once all the exec probes

> in the cluster have a `timeoutSeconds` value set.

Reviewer: We should document that this feature gate won't exist forever and they it will be locked to true at some point

try `failureThreshold` times before giving up. Giving up in case of liveness probe means restarting the container. In case of readiness probe the Pod will be marked Unready.
Defaults to 3. Minimum value is 1.

{{< note >}}
Before Kubernetes 1.20, the field `timeoutSeconds` was not respected for exec probes:
probes continued running indefinitely, even past their configured deadline,
until a result was returned.

This defect was corrected in Kubernetes v1.20. You may have been relying on the previous behavior,
even without realizing it, as the default timeout is 1 second.

> As a cluster administrator, you can disable the feature gate `ExecProbeTimeout` (set it to `false`)

Reviewer 2: nit:
```suggestion
As a cluster administrator, you can disable the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) `ExecProbeTimeout` (set it to `false`)
```

> on kubelet to restore the  behavior from older versions, then remove that override

Reviewer 2: ```suggestion
on each kubelet to restore the  behavior from older versions, then remove that override
```

> once all the exec probes in the cluster have a `timeoutSeconds` value set.

Reviewer: Future versions of Kubernetes will lock the `ExecProbeTimeout` feature gate to `true`, pods that are impacted from the default 1 second timeout should have their probe timeouts updated to an appropriate value.

Reviewer 2: How about:
```suggestion
If you have pods that are impacted from the default 1 second timeout,
you should update their probe timeout so that you're ready for the
eventual removal of that feature gate.
```

If there's a defined version when it will be removed, and that's a firm commitment, let's document that detail.
